### ISSUE TITLE
Assigning to free variables in the node environment should not trigger no-global-assign errors.

### ISSUE THREAD

**Issue by jacksonrayhamilton** [eslint/eslint#9445]
**Tell us about your environment**

* **ESLint Version:** 4.9.0
* **Node Version:** 6.11.4
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?** Default.

**Please show your full configuration:**

```js
module.exports = {
  env: {
    node: true,
  },
  rules: {
    'no-global-assign': ['error'],
  },
};
```

**What did you do? Please include the actual source code causing the issue.**

```js
require = function () {};
```

**What did you expect to happen?** `require` is a free variable in Node, it's not a global variable.  (Try evaluating `'require' in global` in a JavaScript file in Node.  It evaluates to `false`.)  Therefore, no error should be reported.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint reports the following error:

```
1:1  error  Read-only global 'require' should not be modified  no-global-assign
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. This is a result of the abstraction that ESLint uses to expose globals (see [here](https://eslint.org/docs/user-guide/configuring#specifying-globals)). A "global", as far as ESLint is concerned, is just a variable that is statically assumed to be externally defined for a file. This *usually* corresponds to a property on the global object at runtime, but as you've pointed out, it's not always the same thing (Node wraps modules in a function, and variables like `require` are actually parameters, not globals). So I think this is working as designed.

Is there a reason you're trying to assign to `require`? My impression is that in almost all cases, the distinction between ESLint's globals and runtime globals doesn't make a difference in practice as far as best practices are concerned. For example, it generally doesn't make sense to assign to `require`, even though it doesn't actually mutate the global object.

**Comment by jacksonrayhamilton** [comment]
I'm assigning to it when redefining `require`, as suggested [here](https://github.com/standard-things/esm#getting-started):

```js
require = require("@std/esm")(module)
module.exports = require("./main.mjs").default
```

**Comment by jacksonrayhamilton** [comment]
Given your definition of "externally defined" that _might_ make this behavior excusable.  Although I think what I've given is a legitimate and relatively safe use case.  Assigning to `exports` in the `module.exports = exports = {}` fashion also seems legitimate.

**Comment by platinumazure** [comment]
@jacksonrayhamilton I think this might be a case where you should use [eslint-disable comments](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments). We can't cover every corner case and on rare occasions, a rule just might report something that is either contrary to more important requirements in your codebase, or outright incorrect.

I don't think we can make a change here without having ESLint know too much about CommonJS scoping rules, which we don't really want to do (for the purpose of trying to keep the library runtime-agnostic).

**Comment by jacksonrayhamilton** [comment]
Fair enough, especially considering CommonJS always requires some preprocessing.  ESLint can't possibly anticipate the infinitely many ways in which legitimate JS code could be preprocessed and how that could thus affect the semantics on the code it's analyzing.  CommonJS may be popular, but the "magic" involved doesn't lend itself well to generic static analysis.
